Mathlib/Analysis/Complex/HasPrimitives.lean

```diff
@@ -0,0 +1,491 @@
+  rw [Asymptotics.IsLittleO] at this ⊢
+  intro ε ε_pos
+  have := this ε_pos
+  simp only [Asymptotics.isBigOWith_iff, Pi.one_apply, norm_one, mul_one ] at this ⊢
+  have : ∀ᶠ (w : ℂ) in 𝓝 z, ∀ y ∈ Ι z.im w.im, ‖f (w.re + y * I) - f z‖ ≤ ε
+  · rw [Metric.nhds_basis_closedBall.eventually_iff] at this ⊢
+    obtain ⟨i, i_pos, hi⟩ := this
+    refine ⟨i, i_pos, fun w w_in_ball y y_in_I ↦ hi (mem_closedBall_aux w_in_ball y_in_I)⟩
+  apply this.mono (fun w hw ↦ ?_)
+  calc
+    _ ≤ ε * |w.im - z.im|  := intervalIntegral.norm_integral_le_of_norm_le_const hw
+    _ = ε * |(w - z).im| := by simp
+    _ ≤ ε  * ‖w - z‖ := by gcongr; apply abs_im_le_abs
+
+/--   The vertical integral of `f` from `w.re + z.im * I` to `w` is equal to `(w - z).im * f z`
+  up to `o(w - z)`, as `w` tends to `z`. -/
+lemma deriv_of_wedgeInt_im : (fun w ↦ (∫ y in z.im..w.im, f (w.re + y * I)) - (w - z).im * f z)
+    =o[𝓝 z] fun w ↦ w - z := by
+  calc
+    _ = (fun w:ℂ ↦ (∫ y in z.im..w.im, f (w.re + y * I)) - (∫ _ in z.im..w.im, f z)) :=
+      by congr! 2; simp
+    _ =ᶠ[𝓝 z] (fun w ↦ ∫ y in z.im..w.im, f (w.re + y * I) - f z) := ?_
+    _ =o[𝓝 z] fun w ↦ w - z := deriv_of_wedgeInt_im' f_cont hz
+  let r₁ := r - dist z c
+  have : 0 < r₁ := by simp only [mem_ball, gt_iff_lt] at hz ⊢; linarith
+  filter_upwards [ball_mem_nhds z this]
+  intro w hw
+  rw [intervalIntegral.integral_sub ?_ continuousOn_const.intervalIntegrable]
+  exact (f_cont.im_aux_1 hw).intervalIntegrable
+
+/-- The `WedgeInt` has derivative at `z` equal to `f z`. -/
+theorem deriv_of_wedgeInt (hf : VanishesOnRectanglesInDisc c r f) :
+    HasDerivAt (fun w ↦ WedgeInt c w f) (f z) z := by
+  dsimp [HasDerivAt, HasDerivAtFilter]
+  rw [hasFDerivAtFilter_iff_isLittleO]
+  calc
+    _ =ᶠ[𝓝 z] (fun w ↦ WedgeInt z w f - (w - z) * f z) := ?_
+    _ = (fun w ↦ (∫ x in z.re..w.re, f (x + z.im * I)) - (w - z).re * f z)
+        + I • (fun w ↦ (∫ y in z.im..w.im, f (w.re + y * I)) - (w - z).im * f z) := ?_
+    _ =o[𝓝 z] fun w ↦ w - z :=
+      (deriv_of_wedgeInt_re f_cont hz).add ((deriv_of_wedgeInt_im f_cont hz).const_smul_left I)
+  · filter_upwards [VanishesOnRectanglesInDisc.diff_of_wedges f_cont hz hf]
+    exact fun _ ha ↦ by rw [ha]; congr
+  ext1 w
+  simp only [WedgeInt, smul_eq_mul, sub_re, ofReal_sub, sub_im, Pi.add_apply, Pi.smul_apply]
+  set intI := ∫ (x : ℝ) in z.re..w.re, f (x + z.im * I)
+  set intII := ∫ (y : ℝ) in z.im..w.im, f (w.re + y * I)
+  calc
+    _ = intI + I * intII - ((w - z).re + (w - z).im * I) * f z := by congr; rw [re_add_im]
+    _ = intI + I * intII - ((w.re - z.re) + (w.im - z.im) * I) * f z := by simp
+    _ = intI - (w.re - z.re) * f z + I * (intII - (w.im - z.im) * f z) := by ring
+
+end WedgeIntDeriv
+
+/-- *** Moreira's theorem *** A function which is continuous on a disc and whose integral on
+  rectangles in the disc vanishes has a primitive on the disc. -/
+theorem moreiras_theorem {c : ℂ} {r : ℝ} {f : ℂ → ℂ} (f_cont : ContinuousOn f (ball c r))
```

Comment:
I wonder if this should be given a name reflecting its content, rather than its discoverer? This is the general mathlib convention – e.g., to pick a random example, the Hahn-Banach theorem is called `exists_extension_norm_eq`, not `hahn_banach_theorem`.